File: utils/ALU.py
from utils.flag import OVERFLOW, UNDERFLOW, ZERO

class ALU:
    def __init__(self, a_reg, b_reg, f_reg):
        self._a_reg = a_reg
        self._b_reg = b_reg
        self._f_reg = f_reg
    
    def add(self):
        self._f_reg.reset()
        res = self._b_reg.set(self._a_reg.get()+self._b_reg.get())
        if res < 0:
            self._f_reg.setFlag(UNDERFLOW)
        elif res > 0:
            self._f_reg.setFlag(OVERFLOW)
        elif self._b_reg.get() == 0:
            self._f_reg.setFlag(ZERO)

    def sub(self):
        self._f_reg.reset()
        res = self._b_reg.set(self._a_reg.get()-self._b_reg.get())
        if res < 0:
            self._f_reg.setFlag(UNDERFLOW)
        elif res > 0:
            self._f_reg.setFlag(OVERFLOW)
        elif self._b_reg.get() == 0:
            self._f_reg.setFlag(ZERO)

File: armes_cpu_lib/src/ALU.rs
pub struct ALU {}

use ConnectedRegister;
use flags;

impl ALU {
    pub fn add(a_reg: &ConnectedRegister, b_reg: &mut ConnectedRegister, f_reg: &mut flags::F_Register) {
        let b = b_reg._get();
        b_reg._set(a_reg._get() + b);
        if (b_reg._get() != a_reg._get() + b){
            f_reg.set(flags::F_OVERFLOW);
        }else{
            f_reg.unset(flags::F_OVERFLOW);
        }
    }

    pub fn sub(a_reg: &ConnectedRegister, b_reg: &mut ConnectedRegister, f_reg: &mut flags::F_Register) {
        let b = b_reg._get();
        b_reg._set(a_reg._get() - b);
        if (b_reg._get() == 0){
            f_reg.set(flags::F_ZERO);
        }else{
            f_reg.unset(flags::F_ZERO);
        }
    }
}